UI/LobbyMapUI.cs

```diff
@@ -0,0 +1,899 @@
+                }
+            }
+
+            // sort lobbies by their index
+            lobbySelections.Sort((lhs, rhs) => lhs.Info.LobbyIndex - rhs.Info.LobbyIndex);
+
+            // select the current lobby
+            selectedLobbyIndex = lobbySelections.FindIndex(s => s.SID == level.Session.Area.SID);
+            selectedWarpIndexes = new int[lobbySelections.Count];
+        }
+
+        /// <summary>
+        /// Configures the UI for the currently selected lobby.
+        /// </summary>
+        public void updateSelectedLobby(bool first = false) {
+            var selection = lobbySelections[selectedLobbyIndex];
+            var markers = lobbySelections[selectedLobbyIndex].Markers;
+            lobbyMapInfo = selection.Info;
+
+            // get or create a visit manager
+            visitManager = new LobbyVisitManager(selection.SID, selection.Room);
+
+            // generate the 2d array of visited tiles
+            visitedTiles = generateVisitedTiles(lobbyMapInfo, visitManager);
+            
+            // find warps
+            allWarps.Clear();
+            activeWarps.Clear();
+            allWarps.AddRange(markers.Where(f => f.Type == LobbyMapController.MarkerType.Warp).OrderBy(f => f.MarkerId));
+            activeWarps.AddRange(openedWithRevealMap ? allWarps : allWarps.Where(w => isVisited(w.Position)));
+
+            // regenerate marker components
+            markerComponents.ForEach(c => c.RemoveSelf());
+            markerComponents.Clear();
+            markerComponents.AddRange(markers.Where(f => lobbyMapInfo.ShouldShowMarker(f)).Select(createMarkerComponent));
+            markerComponents.ForEach(Add);
+            
+            // if this is the first time we've selected a lobby, select the nearest warp
+            if (first && Engine.Scene is Level level && level.Tracker.GetEntity<Player>() is Player player) {
+                selectedWarpIndexes[selectedLobbyIndex] = 0;
+                var nearestWarpLengthSquared = float.MaxValue;
+                for (int i = 0; i < activeWarps.Count; i++) {
+                    var warpPosition = activeWarps[i].Position + level.LevelOffset;
+                    var lengthSquared = (warpPosition - player.Position).LengthSquared();
+                    if (lengthSquared < nearestWarpLengthSquared) {
+                        nearestWarpLengthSquared = lengthSquared;
+                        selectedWarpIndexes[selectedLobbyIndex] = i;
+                    }
+                }
+            }
+
+            var warpIndex = selectedWarpIndexes[selectedLobbyIndex];
+            
+            // get the map texture
+            mapTexture = GFX.Gui[lobbyMapInfo.MapTexture].Texture.Texture;
+
+            // generate the overlay texture
+            overlayTexture?.Dispose();
+            overlayTexture = new Texture2D(Engine.Instance.GraphicsDevice, lobbyMapInfo.RoomWidth, lobbyMapInfo.RoomHeight, false, SurfaceFormat.Alpha8);
```

Comment:
I doubt people will use this in rooms of absurd size but if they do you'll have a problem here lol. dimensions bigger than 4096 will crash on XNA. you could fix by capping it and scaling both dimensions down but idk at that point maybe they deserve to crash

Reply:
Block XNA in Everest installations tbh. ;)